File: Codigos estudiantes por lenguaje/PY/Luciano_Chavarria/Matemáticas/Criba_Optimizada.py
from sys import stdout, stdin

wr = stdout.write

MAX_SIZE = 1000001

isprime = [True for x in range(MAX_SIZE)]

prime = []

SPF = [int for x in range(MAX_SIZE)]

def manipuledSieve(N):
    global isprime
    global prime
    global SPF
    isprime[0] = False
    isprime[1] = False
    for i in range(2,N):
        if isprime[i]:
            prime.append(i)
            SPF[i] = i
        j = 0
        while j < len(prime) and i * prime[j] < N and prime[j] <= SPF[i]:
            isprime[i*prime[j]] = False
            SPF[i*prime[j]] = prime[j]
            j+=1

N = 100
for i in range(MAX_SIZE):
    # isprime.append(True)
    SPF.append(2)
manipuledSieve(N)
for i in range(len(prime)):
    wr(f"{prime[i]}  ")

wr("\n")

for i in range(2,N):
    wr(f"{i} : {SPF[i]} \n")

File: Codigos estudiantes por lenguaje/C++/Daniel Cangrejo/Matematicas/Criba_Optimizada.cpp
#include <iostream>
#include <vector>

using namespace std;

typedef long long int lli;

const lli MAX_SIZE = 102400;

vector<lli> isPrime(MAX_SIZE, true);
vector<lli> prime;
vector<lli> SPF(MAX_SIZE);
vector<lli> hexaPrime;

void manipuledSieve(int N){
    isPrime[0] = isPrime[1] = false;
    for(lli i = 2; i < N; i++){
        if(isPrime[i]){
            prime.emplace_back(i);
            SPF[i] = i;
        }
        for(lli j = 0; j < (lli)prime.size() && i * prime[j] < N && prime[j] <= SPF[i]; j++){
            isPrime[i*prime[j]] = false;
            SPF[i*prime[j]] = prime[j];
        }
    }
}

int main()
{
    ios_base::sync_with_stdio(false);cin.tie(NULL);cout.tie(NULL);
    lli N; cin>>N;
    manipuledSieve(N);
    for(lli i = 0; i < (lli)prime.size() && prime[i] <= N; i++){
        cout<<prime[i]<<" ";
    }
    cout<<endl;
    return 0;
}
